cli/util/uploader.go

```diff
@@ -0,0 +1,434 @@
+	"github.com/nspcc-dev/neofs-sdk-go/user"
+	"github.com/nspcc-dev/neofs-sdk-go/version"
+	"github.com/urfave/cli/v2"
+)
+
+const (
+	searchBatchSize     = 10000       // Number of objects to search in a batch for finding max block in container.
+	maxParallelSearches = 40          // Control the number of concurrent searches.
+	oidSize             = sha256.Size // Size of object ID.
+
+	numWorkers = 100 // Number of workers to fetch and upload blocks concurrently.
+)
+
+func uploadBin(ctx *cli.Context) error {
+	if err := cmdargs.EnsureNone(ctx); err != nil {
+		return err
+	}
+	rpcNeoFS := ctx.StringSlice("fs-rpc-endpoint")
+	containerIDStr := ctx.String("container")
+	attr := ctx.String("block-attribute")
+	acc, _, err := options.GetAccFromContext(ctx)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to load account: %v", err), 1)
+	}
+
+	var containerID cid.ID
+	if err = containerID.DecodeString(containerIDStr); err != nil {
+		return cli.Exit(fmt.Sprintf("failed to decode container ID: %v", err), 1)
+	}
+
+	gctx, cancel := options.GetTimeoutContext(ctx)
+	defer cancel()
+	rpc, err := options.GetRPCClient(gctx, ctx)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to create RPC client: %v", err), 1)
+	}
+
+	currentBlockHeight, err := rpc.GetBlockCount()
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to get current block height from RPC: %v", err), 1)
+	}
+	fmt.Fprintln(ctx.App.Writer, "Chain block height:", currentBlockHeight)
+
+	signer := user.NewAutoIDSignerRFC6979(acc.PrivateKey().PrivateKey)
+	p, err := pool.New(pool.NewFlatNodeParams(rpcNeoFS), signer, pool.DefaultOptions())
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to create NeoFS pool: %v", err), 1)
+	}
+
+	if err = p.Dial(context.Background()); err != nil {
+		return cli.Exit(fmt.Sprintf("failed to dial NeoFS pool: %v", err), 1)
+	}
+	defer p.Close()
+	net, err := p.NetworkInfo(ctx.Context, client.PrmNetworkInfo{})
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to get network info: %w", err), 1)
+	}
+	homomorphicHashingDisabled := net.HomomorphicHashingDisabled()
+	lastMissingBlockIndex, err := fetchLatestMissingBlockIndex(ctx.Context, p, containerID, acc.PrivateKey(), attr, int(currentBlockHeight))
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to fetch max block index from container: %w", err), 1)
+	}
+
+	fmt.Fprintln(ctx.App.Writer, "First block of latest incomplete batch uploaded to NeoFS container:", lastMissingBlockIndex)
+
+	if lastMissingBlockIndex > int(currentBlockHeight) {
+		fmt.Fprintf(ctx.App.Writer, "No new blocks to upload. Need to upload starting from %d, current height %d\n", lastMissingBlockIndex, currentBlockHeight)
+		return nil
+	}
+
+	for batchStart := lastMissingBlockIndex; batchStart <= int(currentBlockHeight); batchStart += searchBatchSize {
+		batchEnd := min(batchStart+searchBatchSize, int(currentBlockHeight)+1)
+		fmt.Fprintf(ctx.App.Writer, "Processing batch from %d to %d\n", batchStart, batchEnd)
```

Comment:
s/batchEnd/betchEnd-1

Because block number `batchEnd` is processed in the next batch.